### DISCUSSION TITLE
Wolfenstein Blog - Part 1 - Question

### DISCUSSION THREAD

**Discussion by maptz** [JamesRandall/csharp-wolfenstein#2] [General]
Hi James, 

I was reading your Wolfenstein c# porting blog and decided I'd try looking at the endering code you'd implemented. 

Following Part 1, I tried implementing identical code in WPF and Blazor Webassembly to see what the differences were. 

To that end, I updated your `Renderer` class so that it showed a different image on every frame:

```csharp
public uint[,] UpdateFrameBuffer()
	{

		for (int row = 0; row < _height; row++)
			for (int col = 0; col < _width; col++)
			{
				var r = (byte)_random.Next(_height);
				var b = (byte)_random.Next(_width);
				_buffer[row, col] = MakePixel(r, 0, b, 0xFF);
			}
		return _buffer;
	}
```

For some reason, the FPS went right down to 1fps.  Are you seeing a similar effect? Have you got to the point when you're updating the actual pixels on the screen through this method? Am I missing something that you've implemented further down the chain?

**Comment by JamesRandall** [comment]
I update the frame buffer in quite a different way - I'm using unsafe code and pointers rather than accessing the array elements through the index property. I've written a bit about this in part 5:

https://www.jamesdrandall.com/posts/csharp_blazor_wolfenstein_part_5_decoding_maps_and_performance/

Part 6 is on the way and its fully rendering walls, doors and ceilings now and runs at about 30fps in a debug build and 40fps in a release build.

The array index approach was definitely slower but I don't recall it being that slow. I think I was running at around 20fps. in a release build.

Be interested to see what switching to pointers does, if its still running like a snail then pop the code somewhere and I'll take a look.

**Reply by maptz** [reply]
Thanks for the response @JamesRandall.

FYI the displayed FPS only went low when I added the Randomisation code. I was suspicious that the canvas wasn't actually being updated at the FPS shown in the part 1 code. I thought the displayed FPS might be just a snapshot of a recent FPS - followed by a period with no canvas update - if you see what I mean. 

I'll take a look at the later parts on your blog and see what you're doing now.  I should have continued reading before ploughing headlong into the code :)
